diff --git a/node_modules/@material-ui/core/ButtonBase/ButtonBase.js b/node_modules/@material-ui/core/ButtonBase/ButtonBase.js
index 4972129..c121c31 100644
--- a/node_modules/@material-ui/core/ButtonBase/ButtonBase.js
+++ b/node_modules/@material-ui/core/ButtonBase/ButtonBase.js
@@ -159,7 +159,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
     };
   }, []);
   React.useEffect(function () {
-    if (focusVisible && focusRipple && !disableRipple) {
+    if (focusVisible && focusRipple && !disableRipple && rippleRef.current) {
       rippleRef.current.pulsate();
     }
   }, [disableRipple, focusRipple, focusVisible]);
@@ -241,7 +241,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
       keydownRef.current = true;
       event.persist();
       rippleRef.current.stop(event, function () {
-        rippleRef.current.start(event);
+        rippleRef.current && rippleRef.current.start(event);
       });
     }
 
@@ -269,7 +269,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
       keydownRef.current = false;
       event.persist();
       rippleRef.current.stop(event, function () {
-        rippleRef.current.pulsate(event);
+        rippleRef.current && rippleRef.current.pulsate(event);
       });
     }
 
diff --git a/node_modules/@material-ui/core/esm/ButtonBase/ButtonBase.js b/node_modules/@material-ui/core/esm/ButtonBase/ButtonBase.js
index 58cfc64..1ef03e2 100644
--- a/node_modules/@material-ui/core/esm/ButtonBase/ButtonBase.js
+++ b/node_modules/@material-ui/core/esm/ButtonBase/ButtonBase.js
@@ -135,7 +135,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
     };
   }, []);
   React.useEffect(function () {
-    if (focusVisible && focusRipple && !disableRipple) {
+    if (focusVisible && focusRipple && !disableRipple && rippleRef.current) {
       rippleRef.current.pulsate();
     }
   }, [disableRipple, focusRipple, focusVisible]);
@@ -217,7 +217,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
       keydownRef.current = true;
       event.persist();
       rippleRef.current.stop(event, function () {
-        rippleRef.current.start(event);
+        rippleRef.current && rippleRef.current.start(event);
       });
     }
 
@@ -245,7 +245,7 @@ var ButtonBase = /*#__PURE__*/React.forwardRef(function ButtonBase(props, ref) {
       keydownRef.current = false;
       event.persist();
       rippleRef.current.stop(event, function () {
-        rippleRef.current.pulsate(event);
+        rippleRef.current && rippleRef.current.pulsate(event);
       });
     }
 
